### PR TITLE
new:dev:SDK-386_improvement: Changing retry conditions for 503 response

### PR DESCRIPTION
**Pull request by shekharsaurabh**
Making 503 response code a retryable exception for all HTTP requests.

### CODE REVIEW

File: qds_sdk/exception.py [hunk 1]
```diff
@@ -88,6 +88,7 @@ class MethodNotAllowed(ClientError):
 
 
 class ApiThrottledRetry(ClientError):
-    """An error raised when upstream requests are throttled."""
+    """An error raised when upstream requests are throttled. or Service Unavailable"""
```
Comment (chattarajoy) [inline qds_sdk/exception.py]:
```suggestion
    """An error raised when upstream requests are throttled or Service Unavailable"""
```

File: qds_sdk/connection.py [hunk 2]
```diff
@@ -188,15 +188,18 @@ def _handle_error(response):
...
             sys.stderr.write(response.text + "\n")
             raise RetryWithDelay(response, "Data requested is unavailable. Retrying...")
-        elif code == 429:
+        elif code in (429, 503):
             sys.stderr.write(response.text + "\n")
-            raise ApiThrottledRetry(response, "Too many requests. Retrying...")
+            if code == 429
+              raise ApiThrottledRetry(response, "Too many requests. Retrying...")
+            else
+              raise ApiThrottledRetry(response, "Service Unavailable. Retrying...")
```
Comment (chattarajoy) [inline qds_sdk/connection.py]: Define a common method `get_error_message` to get the error message
```suggestion
            raise ApiThrottledRetry(get_error_message(code))
```

File: qds_sdk/connection.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def _api_call_raw(self, req_type, path, data=None, params=None):
         else:
             raise NotImplemented
 
-        self._handle_error(r)
+        self._handle_error(self, r)
```
Comment (chattarajoy) [inline qds_sdk/connection.py]: Remove static method from method definition, no need to pass self explicitly

File: qds_sdk/connection.py [hunk 4]
```diff
@@ -188,15 +188,15 @@ def _handle_error(response):
...
             raise RetryWithDelay(response)
         elif code == 449:
             sys.stderr.write(response.text + "\n")
-            raise RetryWithDelay(response, "Data requested is unavailable. Retrying...")
-        elif code == 429:
+            raise RetryWithDelay(response, self.get_error_message(code))
+        elif code in (429, 503):
             sys.stderr.write(response.text + "\n")
-            raise ApiThrottledRetry(response, "Too many requests. Retrying...")
+            raise ApiThrottledRetry(response, self.get_error_message(code))
```
Comment (chattarajoy) [inline qds_sdk/connection.py]: If this is a static method, you don't need to use self here, just call Connection.method_name

Comment (chattarajoy) [reply qds_sdk/connection.py]: No need to change the def of handle error as self is not being used

File: qds_sdk/connection.py [hunk 5]
```diff
@@ -188,15 +188,15 @@ def _handle_error(response):
...
             raise RetryWithDelay(response)
         elif code == 449:
             sys.stderr.write(response.text + "\n")
-            raise RetryWithDelay(response, "Data requested is unavailable. Retrying...")
-        elif code == 429:
-            sys.stderr.write(response.text + "\n")
-            raise ApiThrottledRetry(response, "Too many requests. Retrying...")
+            raise RetryWithDelay(response, Connection._get_error_message(code))
+        elif code in (429, 503):
+            sys.stderr.write("Got %s response." %(str(code))+ response.text + " Retrying..." + "\n")
```
Comment (shekharsaurabh) [inline qds_sdk/connection.py]: Tested preview of stderr - 
```sshekhar-mbp:qds-sdk-py sshekhar$ qds.py --token 'a1aff3f6541b4f17857b9d109fb44d65a239d30b15e94f3e8aa8ef135c17f81a' --url "https://master.qa.qubole.net/api" --base_retry_delay=3 --max_retries=2  hivecmd run --query "show tables"
Got 429 response. Retrying...
Got 429 response. Retrying...
Got 429 response. Retrying...
Error: Status code 429 (ApiThrottledRetry) from url https://master.qa.qubole.net/api/v1.2/commands```

File: qds_sdk/connection.py [hunk 6]
```diff
@@ -78,23 +78,23 @@ def f_retry(self, *args, **kwargs):
             return f_retry  # true decorator
         return deco_retry
 
-    @retry((RetryWithDelay, requests.Timeout, ServerError, ApiThrottledRetry))
```
Comment (chattarajoy) [inline qds_sdk/connection.py]: @shekharsaurabh let's not change the names for existing exception classes. It has been there for a long time and there might be users who are specifically handling these exception classes. RetryWithDelay should not be changed

File: qds_sdk/connection.py [hunk 7]
```diff
@@ -188,15 +188,18 @@ def _handle_error(response):
...
             raise RetryWithDelay(response)
         elif code == 449:
             sys.stderr.write(response.text + "\n")
             raise RetryWithDelay(response, "Data requested is unavailable. Retrying...")
-        elif code == 429:
+        elif code in (429, 503):
             sys.stderr.write(response.text + "\n")
-            raise ApiThrottledRetry(response, "Too many requests. Retrying...")
+            if code == 429
+              raise ApiThrottledRetry(response, "Too many requests. Retrying...")
```
[the remark this answers is not in the document]
Comment (chattarajoy) [reply qds_sdk/connection.py]: yeah, we can, just the messages can be different.

**Pull request merged**